### PR TITLE
[ML] Allow datafeed and job configs for datafeed preview API

### LINKED ISSUE #70264: [ML] Datafeed _preview should work with configs
The anomaly detection datafeed `_preview` currently requires that the datafeed has already been created. https://www.elastic.co/guide/en/elasticsearch/reference/7.11/ml-preview-datafeed.html

As an enhancement, it would be useful to preview source data before creating it. For example:
- To check if relevant data exists before creating the job - avoiding unnecessary object creation, unnecessary memory usage and in order to explain to users why no anomalies are being found
- To check that the expected data will be analysed - useful if trying to hand craft complex aggs or scripted fields or complex filters
- To simplify UI code which already performs a preview step of its own - will also make sure results match
 
Being able to preview source data by supplying a potential config, also aligns with `POST _ml/data_frame/analytics/_explain` and `POST _transform/_preview`, both of which can be performed prior to creating the job/transform.

### PR DESCRIPTION
**Pull request by benwtrent**
Previously, a datafeed and job must already exist for the `_preview` API to work.

With this change, users can get an accurate preview of the data that will be sent to the anomaly detection job
without creating either of them. 

Example:
```
POST _ml/datafeeds/_preview
{
            "datafeed_config": {
              "indexes":"airline-data"
            },
            "job_config": {
              "analysis_config" : {
                "bucket_span": "1h",
                "detectors" :[{"function":"sum","field_name":"responsetime","by_field_name":"airline"}]
              },
              "data_description" : {
                "time_field":"time"
              }
            }
          }
```

closes https://github.com/elastic/elasticsearch/issues/70264

### CODE REVIEW

File: docs/reference/ml/anomaly-detection/apis/preview-datafeed.asciidoc [hunk 1]
```diff
@@ -43,12 +50,32 @@ supply the credentials.
...
+`datafeed_config`::
+(Optional, object) The {dfeed} definition to preview. For valid definitions, see
+the <<ml-put-datafeed-request-body,create {dfeeds} API>>.
+
+`job_config`::
+(Optional, object) The configuration details for the {anomaly-job} that is
+associated with the {dfeed}. If the `datafeed_config` object does not include a
+`job_id` that references an existing {anomaly-job}, you must supply this
+`job_config` object. If you include both a `job_id` and a `job_config`, the
+latter information is used. For valid definitions, see the
```
Comment (lcawl) [inline docs/reference/ml/anomaly-detection/apis/preview-datafeed.asciidoc:71]: It seems to be implied that you can't supply *only* a job_config. If that's true, I think we should say so explicitly so there's no doubt:
```suggestion
latter information is used. You cannot specify a `job_config` object unless you also supply a `datafeed_config` object. For valid definitions, see the
```

File: client/rest-high-level/src/test/java/org/elasticsearch/client/ml/PreviewDatafeedRequestTests.java [hunk 2]
```diff
@@ -27,6 +30,6 @@ protected PreviewDatafeedRequest doParseInstance(XContentParser parser) throws I
 
     @Override
     protected boolean supportsUnknownFields() {
-        return true;
+        return false;
```
Comment (davidkyle) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/ml/PreviewDatafeedRequestTests.java:33]: ++ requests parsering shouldn't be lenient. Can you also change the parser in `PreviewDatafeedRequest` to have `ignoreUnknownFields = false`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PreviewDatafeedAction.java [hunk 3]
```diff
@@ -87,7 +143,53 @@ public boolean equals(Object obj) {
...
+                if (datafeedBuilder != null) {
+                    datafeedBuilder.setId("preview_id");
+                    if (datafeedBuilder.getJobId() == null && jobBuilder == null) {
+                        throw new IllegalArgumentException("[datafeed_config.job_id] must be set or a [job_config] must be provided");
+                    }
+                    if (datafeedBuilder.getJobId() == null) {
+                        datafeedBuilder.setJobId("preview_job_id");
+                    }
+                }
+                if (jobBuilder != null) {
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PreviewDatafeedAction.java:181]: Is it valid to have a jobBuilder but no datafeedBuilder a null datafeedId? i.e. just the job config

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPreviewDatafeedAction.java [hunk 4]
```diff
@@ -65,34 +67,50 @@ public TransportPreviewDatafeedAction(Settings settings, ThreadPool threadPool,
...
-                        });
-                    },
+                    jobBuilder -> previewDatafeed(datafeedConfig, jobBuilder.build(), listener),
                     listener::onFailure));
             },
-            listener::onFailure));
+            listener::onFailure
+        );
+        if (request.getDatafeedConfig() != null) {
+            datafeedConfigActionListener.onResponse(request.getDatafeedConfig());
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPreviewDatafeedAction.java:83]: Well if I'd known it was that simple...

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestPreviewDatafeedAction.java [hunk 5]
```diff
@@ -35,8 +39,10 @@ public String getName() {
 
     @Override
     protected RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient client) throws IOException {
-        String datafeedId = restRequest.param(DatafeedConfig.ID.getPreferredName());
-        PreviewDatafeedAction.Request request = new PreviewDatafeedAction.Request(datafeedId);
+        String datafeedId = restRequest.param(DatafeedConfig.ID.getPreferredName(), null);
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestPreviewDatafeedAction.java:42]: The `defaultValue` of `null` is unnecessary but I prefer this usage it makes the intent more explicit

**Review by davidkyle — APPROVED** [review]
LGTM

Nice, I thought this would be a lot more complicated

File: client/rest-high-level/src/test/java/org/elasticsearch/client/ml/PreviewDatafeedRequestTests.java (hunk 2 above)
Comment (benwtrent) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/ml/PreviewDatafeedRequestTests.java:33]: 100%

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PreviewDatafeedAction.java (hunk 3 above)
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PreviewDatafeedAction.java:181]: > Is it valid to have a jobBuilder but no datafeedBuilder a null datafeedId?

no, it isn't. The API MUST have either a datafeedBuilder or a datafeedId. The restrictive ctors ensure this (with null checks).

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestPreviewDatafeedAction.java (hunk 5 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/datafeeds/RestPreviewDatafeedAction.java:42]: Gotcha, I will only grab this param if `hasContent` is false. Also, I won't provide a default value that way it blows up expecting it.

**Comment by benwtrent** [comment]
run elasticsearch-ci/bwc

**Comment by benwtrent** [comment]
run elasticsearch-ci/2

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**